### ISSUE TITLE
Implement WALS matrix factorization in Tensorflow 2.0

### ISSUE THREAD

**Issue by wileeam** [tensorflow/tensorflow#27107]
**System information**
- TensorFlow version (you are using): 2.0-master
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Implement tf.contrib.factorization.WALSMatrixFactorization and tf.contrib.factorization.WALSModel in Tensorflow 2.0 (ideally in a distributed manner)

**Will this change the current api? How?**
Most likely.
The current WALSMatrixFactorization and WALSModel lie under tf.contrib.factorization, which contains some other methods, in particular clustering ones, such as KMeans or GMM.
The factorization itself makes sense under tf.linalg instead (the other methods need further assesment) while the estimator may be somewhere.
In other words, the factorization would probably benefit from being decoupled from the model implementation.

**Who will benefit with this feature?**
Anyone using the current factorization who wants (or needs) to upgrade to Tensorflow 2.0.

**Any Other info.**
There is an open issue related to WALS that could be of interest to take into account if this feature is considered: https://github.com/tensorflow/tensorflow/issues/21991

**Comment by dynamicwebpaige** [comment]
This is an excellent recommendation (no pun intended), and we would welcome the contribution! :slightly_smiling_face: 

If you are looking for a template for WALS implementation, the Google Cloud Platform team has a TensorFlow 1.x example located [here](https://github.com/GoogleCloudPlatform/tensorflow-recommendation-wals). Let us know if you have any questions!

**Comment by SumNeuron** [comment]
@dynamicwebpaige semi-related (also working on the documentation of this), but I have a GSoC proposal for this...

**Comment by SumNeuron** [comment]
@agarwal-ashish  @walidk @dynamicwebpaige  @wileeam any updates?

**Comment by SumNeuron** [comment]
Status?

**Comment by ramanshrivastava** [comment]
Any update?

**Comment by wileeam** [comment]
@SumNeuron I haven't had the time to actually look further into this unfortunately.
On the other hand it seems that NCF is now favoured in Tensorflow 2 instead of WALS for recommendations as stated in the official models repository: https://github.com/tensorflow/models/tree/master/official/recommendation.

However, Tensorflow would benefit from some knowledge sharing from Google after the recent release (in beta though) of this (W)ALS algorithm in their platform (BigQuery ML): https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-create-matrix-factorization
I still believe (W)ALS matrix factorization is relevant, particularly in large-scale scenarios. Although this could open the pandora box of matrix factorization methods.

**Comment by agarwal-ashish** [comment]
Factorization, similar to other contrib packages, is not being maintained as part of core TensorFlow. However I do believe that that WALS is a very important algorithm and I would love to see it being driven by some active community. 
FWIW, @walidk has done extensive studies on WALS and related algorithms, including neural ones, and can shed more light on how well it performs.

**Comment by walidk** [comment]
I agree, WALS remains an important algorithm in recommender systems, and the state of the art for some problems.

On the topic of MLP-learned similarity (such as NCF) Vs. dot product similarity (such as matrix factorization or two-tower models): I would recommend trying a dot product model by default instead of NCF. It is unclear whether NCF performs as well in practice as dot product models, see for example the following study: https://arxiv.org/pdf/1911.07698.pdf

**Comment by ravgupta161803** [comment]
I have implemented the WALS low rank factorization in tensorflow 2.0 and got some great results from it. Here is the link to the repository https://github.com/ravgupta161803/WALS-matrix-factorisastion-tensorflow-2.0.git

**Comment by sachinprasadhs** [comment]
tf.contrib module has been deprecated, please check [here](https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md) for more details.
If you think this feature needs to be added in the Tensorflow version, please create a new issue. Thanks!
